yt_dlp/downloader/external.py

```diff
@@ -302,6 +316,137 @@ def _make_cmd(self, tmpfilename, info_dict):
             cmd += ['--', info_dict['url']]
         return cmd
 
+    def _call_downloader(self, tmpfilename, info_dict):
+        info_dict.pop('__rpc_port', None)
+        info_dict.pop('__rpc_secret', None)
+        if self._ENABLE_PROGRESS:
+            import uuid
+            info_dict = info_dict.copy()
+            info_dict['__rpc_port'] = find_available_port() or 19190
+            info_dict['__rpc_secret'] = str(uuid.uuid4())
+        return super()._call_downloader(tmpfilename, info_dict)
+
+    def _call_process(self, cmd, info_dict):
+        if '__rpc_port' not in info_dict:
+            return super()._call_process(cmd, info_dict)
+
+        from tempfile import TemporaryFile
+        import json
+        import uuid
+
+        rpc_port = info_dict['__rpc_port']
+        rpc_secret = info_dict['__rpc_secret']
+        nr_frags = len(info_dict['fragments']) if 'fragments' in info_dict else -1
+
+        def aria2c_rpc(method, params, secret=True):
+            if secret and rpc_secret:
+                params = [f'token:{rpc_secret}', *params]
+            # note: there's no need to be UUID (it can even a numeric value), but that's easier
+            sanitycheck = str(uuid.uuid4())
+            d = json.dumps({
+                'jsonrpc': '2.0',
+                'id': sanitycheck,
+                'method': method,
+                'params': params,
+            }).encode('utf-8')
+            request = sanitized_Request(
+                f'http://localhost:{rpc_port}/jsonrpc',
+                headers={
+                    'Content-Type': 'application/json',
+                    'Content-Length': f'{len(d)}',
+                    'Ytdl-request-proxy': '__noproxy__',
+                },
+                data=d)
+            with self.ydl.urlopen(request) as r:
+                resp = json.load(r)
+            # failing at this assertion means that the RPC server went wrong
+            # (KeyEror includes)
+            assert resp['id'] == sanitycheck
+            return resp['result']
+
+        started = time.time()
+        status = {
+            'filename': info_dict.get('_filename'),
+            'status': 'downloading',
+            'elapsed': 0,
+            'downloaded_bytes': 0,
+        }
+        if nr_frags >= 0:
+            status.update({
+                'fragment_count': nr_frags,
+                'fragment_index': 0,
+            })
+        self._hook_progress(status, info_dict)
+
+        with TemporaryFile() as so, TemporaryFile() as se, \
```

Comment:
why use tempfile? Cant `subprocess.PIPE` be used?

Reply:
I remember there were something messy when they're PIPE, or I would have use it
Since pipes has capacity, it has to keep reading to keep aria2c running
But I don't think Python has ability to check how many bytes can be read

Reply:
We don't actually need stdout at all, and I don't think stderr will fill up

Reply:
okay let's DEVNULL them